src/test/java/org/mabrarov/exceptionsafety/FactoryMethodTest.java

```diff
@@ -32,193 +33,212 @@
-  public void test_closeCheckedException() {
-    resourceConfigurator = resourceConfiguratorNoException;
-    resourceCreator = resourceCreatorCloseCheckedException;
-    verifyResourceCreationAndClosing();
-  }
-
-  @Test
-  public void test_closeRuntimeException() {
-    resourceConfigurator = resourceConfiguratorNoException;
-    resourceCreator = resourceCreatorCloseRuntimeException;
-    verifyResourceCreationAndClosing();
-  }
-
-  @Test
-  public void test_configurationCheckedException_closeCheckedException() {
-    resourceConfigurator = resourceConfiguratorCheckedException;
-    resourceCreator = resourceCreatorCloseCheckedException;
-    verifyResourceCreationAndClosing();
-  }
-
-  @Test
-  public void test_configurationCheckedException_closeRuntimeException() {
-    resourceConfigurator = resourceConfiguratorCheckedException;
-    resourceCreator = resourceCreatorCloseRuntimeException;
-    verifyResourceCreationAndClosing();
-  }
-
-  @Test
-  public void test_configurationRuntimeException_closeCheckedException() {
-    resourceConfigurator = resourceConfiguratorRuntimeException;
-    resourceCreator = resourceCreatorCloseCheckedException;
-    verifyResourceCreationAndClosing();
-  }
-
-  @Test
-  public void test_configurationRuntimeException_closeRuntimeException() {
-    resourceConfigurator = resourceConfiguratorRuntimeException;
-    resourceCreator = resourceCreatorCloseRuntimeException;
-    verifyResourceCreationAndClosing();
-  }
-
-  @Test
-  public void test_exceptionSuppressionError() {
-    resourceConfigurator = resourceConfiguratorRuntimeExceptionWithSuppressionFailure;
-    resourceCreator = resourceCreatorCloseRuntimeException;
-    verifyResourceCreationAndClosing();
-  }
-
-  private void verifyResourceCreationAndClosing() {
-    try (final OutputStream resource = createConfiguredResource();
-        final PrintStream printStream = new PrintStream(resource)) {
-      Assert.assertFalse("Resource should be opened", resourceClosed.get());
-      printStream.println("Test");
-    } catch (final Exception | TestSuppressionError e) {
-      e.printStackTrace();
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    private final AtomicBoolean resourceClosed = new AtomicBoolean();
+    private ResourceCreator resourceCreator;
+    private ResourceConfigurator resourceConfigurator;
+
+    private OutputStream createConfiguredResource()
+            throws IOException, TestResourceConfigurationException {
+        OutputStream resource = null;
+        Throwable throwable = null;
+        try {
+            resource = createResource();
+            configureResource(resource);
+            return resource;
+        } catch (final IOException | TestResourceConfigurationException | RuntimeException e) {
+            throwable = e;
+            throw e;
+        } catch (final Throwable e) {
+            throwable = e;
+            throw new AssertionError("Should never come here", e);
```

Comment:
What about `java.lang.Error` and its subclasses (like `java.lang.OutOfMemoryError`)? Shouldn't we `throw w` instead of `throw new AssertionError(...)` in such cases?